### ISSUE TITLE
Different quality with init_model & eval_set

### ISSUE THREAD

**Issue by Uptimolli** [catboost/catboost#2824]
Hello, I think i have some bug or missunderstanding in using `init_model` with and without `eval_set` parameter.

First, I split the data into train-test, put the test set into eval_set, and set 5 iterations. I train on the train set, and because of eval_set, it stops at the penultimate(4th) iteration. Then I train the same model with the same eval_set, but this time, I pass the previous model into `init_model`. However, the first iteration of this new model does NOT match the last one (the one that was cut off). This raises a question: although it initially considered the trimmed iteration optimal, now it's producing a different split (judging by the fact that the `learn` value in verbose info differs). But why? I found a similar question on GitHub, but it was about `counter_calc_method`, whereas I don’t have categorical features (I did, but I managed to reproduce this issue with a single numerical feature, so it's unlikely that categorical features are the cause). Moreover, this discrepancy doesn't always occur — it depends on the number of iterations, sample size, and the number of features, making it seem like a bug.

Second, why does the model's performance change when I apply the `.copy()` method to an instance of the model instead of using the instance itself? The only differences I found in the parameters are that some numbers get rounded to the 12th (!) decimal place and `n_features_in_` is reset to zero. But could that really have such a strong impact on how the model trains afterward?

Examples:

Data:
```
import catboost as cb
import pandas as pd

from sklearn.model_selection import train_test_split

df = pd.DataFrame(
    [
        [42500, 5048153],
        [140000, 1753287],
        [32500, 4553920],
        [30000, 3644634],
        [45000, 102706],
        [75000, 1087156],
        [45000, 3782607],
        [90000, 5318073],
        [76500, 4540012],
        [35000, 3009074],
        [86000, 1935592],
        [50000, 2139987],
    ],
    columns=["y", "x"],
)

model_params = {
    "random_state": 42,
    "learning_rate": 0.1,
    # "counter_calc_method": "SkipTest"
}

X, y = df.drop(columns=["y"]), df["y"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```

With `eval_set`:

```
model = cb.CatBoostRegressor(n_estimators=5, **model_params)
model.fit(X_train, y_train, verbose=1, eval_set=(X_test, y_test))

model2 = cb.CatBoostRegressor(n_estimators=5, **model_params)
model2.fit(X_train, y_train, init_model=model, verbose=1, eval_set=(X_test, y_test))

model3 = cb.CatBoostRegressor(n_estimators=10, **model_params)
model3.fit(X_train, y_train, verbose=1, eval_set=(X_test, y_test))

print("In model2, train pred:")
print(model2.predict(X_train))
print("In model3, train pred:")
print(model3.predict(X_train))

# preds are not same
```

Without `eval_set`:

```
model = cb.CatBoostRegressor(n_estimators=5, **model_params)
model.fit(X_train, y_train, verbose=1)

model2 = cb.CatBoostRegressor(n_estimators=5, **model_params)
model2.fit(X_train, y_train, init_model=model, verbose=1)

model3 = cb.CatBoostRegressor(n_estimators=10, **model_params)
model3.fit(X_train, y_train, verbose=1)


print("In model2, train pred:")
print(model2.predict(X_train))
print("In model3, train pred:")
print(model3.predict(X_train))

# now preds are same
```

Without `eval_set`, but with `.copy()` method for instance:
```
model = cb.CatBoostRegressor(n_estimators=5, **model_params)
model.fit(X_train, y_train, verbose=1)

model2 = cb.CatBoostRegressor(n_estimators=5, **model_params)
model2.fit(X_train, y_train, init_model=model.copy(), verbose=1)

model3 = cb.CatBoostRegressor(n_estimators=10, **model_params)
model3.fit(X_train, y_train, verbose=1)

print("In model2, train pred:")
print(model2.predict(X_train))
print("In model3, train pred:")
print(model3.predict(X_train))
```
